### PR TITLE
change to read cni version bump from TKR object as CNI can be user-cu…

### LINKED ISSUE #2429: ClusterBootstrap upgrade after user's initial customization
**Bug description**

For the case where the user had initially provided a customized ClusterBootstrap resource and specified Calico as CNI, [this line in the code](https://github.com/vmware-tanzu/tanzu-framework/blob/14e33df6241f2b79428c1fcfaabd094dad79d159/addons/controllers/clusterbootstrap_controller.go#L470
) should not always copy from the ClusterBootstrapTemplate and instead we might need to have an annotation or label in the ClusterBootstrap object for when the addon was user-customized and if that wasn’t present, then fall back to read from  the template 

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [X] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

### PR DESCRIPTION
**Pull request by maralavi**
Signed-off-by: Marjan Alavi <malavi@vmware.com>

### What this PR does / why we need it
change to read cni version bump from TKR object as CNI can be user-customized

### Which issue(s) this PR fixes

Fixes #2429 

### Describe testing done for PR
Existing unit tests

### Release note

```release-note
package-based-lcm: change to read cni version bump from TKR object as CNI can be user-customized
```

### PR Checklist

- [X] Squash the commits into one or a small number of logical commits
      
- [X] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [X] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: config/crd/bases/run.tanzu.vmware.com_tanzukubernetesreleases.yaml [hunk 1]
```diff
@@ -544,7 +544,7 @@ spec:
             type: object
         type: object
     served: true
-    storage: false
+    storage: true
```
Comment (maralavi) [inline config/crd/bases/run.tanzu.vmware.com_tanzukubernetesreleases.yaml:547]: @imikushin I was not able to create/update spec for TKR ```v1alpha3``` object and @shyaamsn figured out this was the cause. Was it an intentional choice to leave ```storage: true``` for ```v1alpha1``` TKR?

Comment (maralavi) [reply config/crd/bases/run.tanzu.vmware.com_tanzukubernetesreleases.yaml]: Changed to load a local tkr CRD with the above booleans switched and that fixed the problem

File: addons/pkg/util/tkr_util.go [hunk 2]
```diff
@@ -9,27 +9,43 @@ import (
...
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
 	runtanzuv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha1"
+	runtanzuv1alpha3 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
 )
 
-// GetTKRByName gets TKR object given a TKR name
-func GetTKRByName(ctx context.Context, c client.Client, tkrName string) (*runtanzuv1alpha1.TanzuKubernetesRelease, error) {
+// GetTKRByName gets TKR object given a TKR name and a label to help determine version of the TKR object
+func GetTKRByName(ctx context.Context, c client.Client, tkrName, label string) (client.Object, error) {
+	tkrV1Alpha1 := &runtanzuv1alpha1.TanzuKubernetesRelease{}
```
[the remark this answers is not in the document]
Comment (maralavi) [reply addons/pkg/util/tkr_util.go:19]: Sure, can introduce separate methods

File: addons/controllers/clusterbootstrap_controller.go [hunk 3]
```diff
@@ -1374,3 +1381,38 @@ func (r *ClusterBootstrapReconciler) removeAdditionalPackageInstalls(remoteClien
 	}
 	return nil
 }
+
+// getCNIFromTKR tries to find the prefix of the provided CNI RefName in the bootstrap packages of the TKR object associated with the cluster. Upon finding
+// the corresponding bootstrap package name, it returns it as the bumped version of the CNI
+func (r *ClusterBootstrapReconciler) getCNIFromTKR(cni *runtanzuv1alpha3.ClusterBootstrapPackage, cluster *clusterapiv1beta1.Cluster) (string, string, error) {
```
Comment (maralavi) [inline addons/controllers/clusterbootstrap_controller.go:1387]: Moving this to utils to add UT, test coverage will complain otehrwise

File: addons/pkg/util/cluster_util.go [hunk 4]
```diff
@@ -68,7 +67,7 @@ func GetClusterByName(ctx context.Context, c client.Client, namespace, name stri
 }
 
 // GetClustersByTKR gets the clusters using this TKR
-func GetClustersByTKR(ctx context.Context, c client.Client, tkr *runtanzuv1alpha1.TanzuKubernetesRelease) ([]*clusterv1beta1.Cluster, error) {
+func GetClustersByTKR(ctx context.Context, c client.Client, tkr client.Object) ([]*clusterv1beta1.Cluster, error) {
```
Comment (vijaykatam) [inline addons/pkg/util/cluster_util.go:70]: Should this be changed back since you have separate methods now?

Comment (maralavi) [reply addons/pkg/util/cluster_util.go]: Thanks for this. Addressed in latest commit (had to squash before merge with main).

**Review by vijaykatam — APPROVED** [review]
LGTM.

**Pull request merged**